src/tree_builder/actions.rs

```diff
@@ -755,10 +723,39 @@ impl<Handle, Sink> TreeBuilderActions<Handle>
         // FIXME: application cache selection algorithm
     }
 
+    // https://html.spec.whatwg.org/multipage/syntax.html#create-an-element-for-the-token
     fn insert_element(&mut self, push: PushFlag, ns: Namespace, name: Atom, attrs: Vec<Attribute>)
             -> Handle {
-        let elem = self.sink.create_element(QualName::new(ns, name), attrs);
-        self.insert_appropriately(AppendNode(elem.clone()), None);
+        declare_tag_set!(form_associatable =
+            button fieldset input keygen label
+            object output select textarea img);
+
+        declare_tag_set!(reassociatable = form_associatable - img);
+
+        let qname = QualName::new(ns, name);
+        let elem = self.sink.create_element(qname.clone(), attrs.clone());
+
+        let insertion_point = self.appropriate_place_for_insertion(None);
+        let tree_node = match insertion_point {
+            LastChild(ref p) |
+            BeforeSibling(ref p) => p.clone()
+        };
+
+        // Step 4.
+        // TODO: Handle template element case
+        if form_associatable(qname.clone())
+           && self.form_elem.is_some()
+           && !(reassociatable(qname.clone())
+               && attrs.iter().any(|a| a.name == qualname!("","form"))) {
+
+               let form = self.form_elem.as_ref().unwrap().clone();
+               if self.sink.same_home_subtree(tree_node, form.clone()) {
+                   self.sink.associate_with_form(elem.clone(), form)
```

Comment:
Hi @hsivonen, @jdm referred me to you for this particular query: is there a specific reason why the spec requires "associating the form owner" to happen before the insertion of the form control element into the tree? [Spec](https://html.spec.whatwg.org/multipage/syntax.html#creating-and-inserting-nodes:form-associated-element)